### ISSUE TITLE
[Blueprint] "Provider forces requestor to accept or reject results" use case in Concent API

### ISSUE THREAD

**Issue by rwrzesien** [golemfactory/concent#41]

**Comment by rwrzesien** [comment]
@lukasz-glen I have several questions regarding this use case:
1. "Usługa Concent bez komunikacji z Requestorem sprawdza czy komunikat MessageTaskToCompute (zawarty w komunikacie MessageAckReportComputedTask) nie był już wcześniej procesowany." - what does it mean 'procesowany' in this context ?
2. What about scenario when provider sends ForceSubtaskResults for the same task again ? It is not documented, however for previous use cases when we did this scenario, it was also not  documented.
3. What will be in field 'payment_ts' in SubtaskResultsAccepted ?

**Comment by cameel** [comment]
I think I can answer some of these:

> 1) "Usługa Concent bez komunikacji z Requestorem sprawdza czy komunikat MessageTaskToCompute (zawarty w komunikacie MessageAckReportComputedTask) nie był już wcześniej procesowany." - what does it mean 'procesowany' in this context ?

It means that for example if the client sends us `ForceSubtaskResults`, we only accept it if we have never accepted `ForceSubtaskResults` with the same `TaskToCompute` in `send()` before. We'll know that because we store information about all accepted tasks in the database.

> 2) What about scenario when provider sends ForceSubtaskResults for the same task again ? It is not documented, however for previous use cases when we did this scenario, it was also not documented.

The passage you mentioned in 1) answers that.

**Comment by Jakub89** [comment]
According to new changes in documentation, we need to change step 2A. From `ForceSubtaskResultsRejected` to `ServiceRefused`.
```
Usługa Concent bez komunikacji z Requestorem sprawdza czy komunikat MessageTaskToCompute (zawarty w komunikacie MessageAckReportComputedTask) nie był już wcześniej procesowany. Jeśli tak to odpowiada Providerowi komunikatem MessageServiceRefused z podaniem przyczyny i Provider powinien zakończyć komunikację.
```

**Comment by rwrzesien** [comment]
@Jakub89 I have updated blueprints and other related docs.

@lukasz-glen Which `reason` should we use in case 2A ?
Currently there are following values for `reason` field available in ServiceRefused (golem_messages version 1.7.0)
```
    class REASON(enum.Enum):
        TooSmallCommunicationPayment = 'TOO_SMALL_COMMUNICATION_PAYMENT'
        TooSmallRequestorDeposit = 'TOO_SMALL_REQUESTOR_DEPOSIT'
        TooSmallProviderDeposit = 'TOO_SMALL_PROVIDER_DEPOSIT'
        SystemOverloaded = 'SYSTEM_OVERLOADED'
```
Or maybe you will add `TASK_ALREADY_PROCESSED` to ServiceRefused reasons ?

**Comment by shadeofblue** [comment]
@lukasz-glen @rwrzesien indeed, https://github.com/golemfactory/golem-messages/pull/127 awaiting merge

**Comment by rwrzesien** [comment]
Updated on latest changes in golem_messages SubtaskResultsRejected.

**Comment by cameel** [comment]
A few corrections are needed:

1) This use case is supposed to happen after verification so `ForceSubtaskResults` can only be submitted in a time window that starts at `AckReportComputedTask.timestamp` + `SUBTASK_VERIFICATION_TIME` and ends at `AckReportComputedTask.timestamp` + `SUBTASK_VERIFICATION_TIME` + `FORCE_ACCEPTANCE_TIME`.

    We need to add `SUBTASK_VERIFICATION_TIME` setting.

2) In step 2C the inequality has wrong direction. It can never be satisfied

3) Upper time limits in steps 3 an 4B and lower time limit in step 4A should not depend on `ForceSubtaskResults.timestamp`. This is because the time windows must be predictable and the requestor does not really know when exactly provider sends his force message.

    We can safely extend the time window and accept response from the requestor if he sends it early but we must not shorten or move his time window. 

4) `task_id`s in examples should now be strings

5) The right URL for the out of band receive is `http://concent.golem.network/receive-out-of-band/`. I.e. with dashes instead of underscores.

**Comment by rwrzesien** [comment]
@cameel 
1. Could you explain ? I have checked docs and they say:
"Requestor ma obowiązek być podłączonym do sieci w czasie gdy Provider ma czas na odwołanie się do usługi Concent. Czas ten liczony jest od momentu upłynięcia limitu czasu na wysłanie komunikatu przez Requestora (pierwszy limit) do upłynięcia limitu czasu na odwołanie się do usługi Concent przez Providera (drugi limit)."
and the definitions above:
"svt - Subtask Verification Time, limit czasu na weryfikację przez Requestora liczony od stempla czasu komunikatu ReportComputedTask, TODO zdefiniowany w komunikacie lub stała globalna.
fat - Force Acceptance Time, limit czasu na odwołanie się do usługi Concent gdy nie otrzymano komunikatu MessageSubtaskResultsAccepted lub MessageSubtaskResultsRejected, stała globalna."
So it looks like `fat` should be used here.
2. Updated, generally we are aware and we discussed it when implementation was being done.
3.
4. Updated.
5. Updated.

**Comment by cameel** [comment]
1) Please look at the figure titled `Powyższy diagram przedstawia wymuszenie przesłania akceptacji wyników, następuje gdy upłynie czas na weryfikację` on page 20:

    ![image12](https://user-images.githubusercontent.com/137030/36028273-a83e8062-0d9e-11e8-8ccd-7371a68119c8.png)

    "moment upłynięcia limitu czasu na wysłanie komunikatu przez Requestora" refers to the message that accepts or rejects subtask results. It's `AckReportComputedTask.timestamp` + `SUBTASK_VERIFICATION_TIME`.

    "limit czasu na odwołanie się do usługi Concent przez Providera" is `AckReportComputedTask.timestamp` + `SUBTASK_VERIFICATION_TIME` + `FORCE_ACCEPTANCE_TIME`

    > fat - Force Acceptance Time, limit czasu na odwołanie się do usługi Concent **gdy nie otrzymano komunikatu MessageSubtaskResultsAccepted lub MessageSubtaskResultsRejected**, stała globalna."
    > So it looks like fat should be used here.

    The part I marked in bold is important: when the acceptance or rejection message was not received. And the deadline for receiving that message is `AckReportComputedTask.timestamp` + `SUBTASK_VERIFICATION_TIME`.

**Comment by rwrzesien** [comment]
Added info about new setting which will be implemented in https://github.com/golemfactory/concent/issues/106.

**Comment by rwrzesien** [comment]
Update time windows in steps 4A and 4B.

**Comment by rwrzesien** [comment]
@cameel Added.

**Comment by cameel** [comment]
I just looked closer at the messages and I see problems with other examples:
- `timestamp` missing in `report_computed_task` too.
- `timestamp` is not a part of `compute_task_def`. It should be outside.

**Comment by rwrzesien** [comment]
@cameel Updated.
